unumpy/linalg/_multimethods.py

```diff
@@ -0,0 +1,224 @@
+import functools
+import operator
+from uarray import create_multimethod, mark_as, all_of_type, Dispatchable
+import builtins
+
+create_numpy = functools.partial(create_multimethod, domain="numpy.linalg")
+
+from .._multimethods import (
+    ndarray,
+    _identity_argreplacer,
+    _self_argreplacer,
+    _dtype_argreplacer,
+    mark_dtype,
+    _first_argreplacer,
+    _first2argreplacer,
+)
+
+
+@create_numpy(_first_argreplacer)
+@all_of_type(ndarray)
+def multi_dot(arrays):
+    return arrays
+
+
+def matrix_power_default(a, n):
+    if not ndim(a) == 2:
+        raise ValueError("Array must be two-dimensional")
+
+    if not isinstance(n, int):
+        raise ValueError("Exponent must be an integer.")
+
+    if a.shape[0] == a.shape[1]:
+        if n == 0:
+            return eye(a.shape[0], a.shape[1])
+        elif n < 0:
+            a = inv(a)
```

Comment:
You don't need this code at all.

Reply:
Like this?
```python
def matrix_power_default(a, n):
    if n == 0:
        return eye(a.shape[0], a.shape[1])
    elif n < 0:
        a = inv(a)

    eigenvalues, eigenvectors = eig(a)

    diagonal = diag(eigenvalues)

    return multi_dot([eigenvectors, power(diagonal, abs(n)), inv(eigenvectors)])
```

Reply:
Like this:
```
def matrix_power_default(a, n):
    eigenvalues, eigenvectors = eig(a)
    diagonal = diag(eigenvalues)
    return multi_dot([eigenvectors, power(diagonal, n), inv(eigenvectors)])
```

Reply:
Or even this:
```python
def matrix_power_default(a, n):
    eigenvalues, eigenvectors = eig(a)
    return multi_dot([eigenvectors, eigenvalues.reshape((shape(a)[0], 1)), inv(eigenvectors)])
```

Reply:
Referencing the NumPy docs:

> For positive integers n, the power is computed by repeated matrix squarings and matrix multiplications. If `n == 0`, the identity matrix of the same shape as M is returned.
If `n < 0`, the inverse is computed and then raised to the `abs(n)`.

Does this implementation ignore the `n == 0` and `n < 0`  cases then?

Reply:
> Does this implementation ignore the n == 0 and n < 0 cases then?

No, it just uses the properties of matrices. 😄

First, that `diag(power(a, n)) == matrix_power(diag(a), n)` for `ndim(a) == 1`. That is, if `A` is diagonal, then raising it to a power means raising all its diagonal elements to that power.

Second, `diag(a) @ X == a.reshape((shape(a)[0], 1)) * X`.

Reply:
(I edited the implementation a few times to get it right, but it's now okay)

Reply:
You can test it out yourself:

```python
>>> a = np.diag(np.arange(5))
>>> b = np.ones(5)
>>> a @ b
array([[0., 0., 0., 0., 0.],
       [1., 1., 1., 1., 1.],
       [2., 2., 2., 2., 2.],
       [3., 3., 3., 3., 3.],
       [4., 4., 4., 4., 4.]])
>>> b @ a
array([[0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.]])
>>> np.diag(a) * b
array([[0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.]])
>>> np.diag(a).reshape((len(a), 1)) * b
array([[0., 0., 0., 0., 0.],
       [1., 1., 1., 1., 1.],
       [2., 2., 2., 2., 2.],
       [3., 3., 3., 3., 3.],
       [4., 4., 4., 4., 4.]])
```